app/models/spree/gateway/stripe_connect.rb

```diff
@@ -0,0 +1,147 @@
+    end
+
+    def void(response_code, creditcard, gateway_options)
+      provider.void(response_code, {})
+    end
+
+    def cancel(response_code)
+      provider.void(response_code, {})
+    end
+
+    def create_profile(payment)
+      return unless payment.source.gateway_customer_profile_id.nil?
+      options = {
+        email: payment.order.email,
+        login: Stripe.api_key,
+      }.merge! address_for(payment)
+
+      source = update_source!(payment.source)
+      if source.number.blank? && source.gateway_payment_profile_id.present?
+        creditcard = source.gateway_payment_profile_id
+      else
+        creditcard = source
+      end
+
+      response = provider.store(creditcard, options)
+      if response.success?
+        payment.source.update_attributes!({
+          cc_type: payment.source.cc_type, # side-effect of update_source!
+          gateway_customer_profile_id: response.params['id'],
+          gateway_payment_profile_id: response.params['default_source'] || response.params['default_card']
+        })
+
+      else
+        payment.send(:gateway_error, response.message)
+      end
+    end
+
+    private
+
+    # In this gateway, what we call 'secret_key' is the 'login'
+    def options
+      options = super
+      options.merge(:login => Stripe.api_key)
+    end
+
+    def options_for_purchase_or_auth(money, creditcard, gateway_options)
+      options = {}
+      options[:description] = "Spree Order ID: #{gateway_options[:order_id]}"
+      options[:currency] = gateway_options[:currency]
+      options[:stripe_account] = stripe_account_id
+
+      creditcard = token_from_card_profile_ids(creditcard)
+
+      return money, creditcard, options
+    end
+
+    def address_for(payment)
+      {}.tap do |options|
+        if address = payment.order.bill_address
+          options.merge!(address: {
+            address1: address.address1,
+            address2: address.address2,
+            city: address.city,
+            zip: address.zipcode
+          })
+
+          if country = address.country
+            options[:address].merge!(country: country.name)
+          end
+
+          if state = address.state
+            options[:address].merge!(state: state.name)
+          end
+        end
+      end
+    end
+
+    def update_source!(source)
+      source.cc_type = CARD_TYPE_MAPPING[source.cc_type] if CARD_TYPE_MAPPING.include?(source.cc_type)
+      source
+    end
+
+    def token_from_card_profile_ids(creditcard)
+      if token_or_card_id = creditcard.gateway_payment_profile_id
+        if customer = creditcard.gateway_customer_profile_id
+          # Assume the gateway_payment_profile_id is a Stripe card_id
+          # So generate a new token, using the customer_id and card_id
+          return tokenize_instance_customer_card(customer, token_or_card_id)
+        end
+
+        # Assume the gateway_payment_profile_id is a token generated by StripeJS
+        return token_or_card_id
+      end
+    end
```

Comment:
What's your preferred pattern to check whether something exists and then use it if it does?